exp/services/ledgerexporter: Add guide for contributors/developers (#5340)
stellar/go: exp/services/ledgerexporter/DEVELOPER_GUIDE.md

# Ledger Exporter Developer Guide
The ledger exporter is a tool to export Stellar network transaction data to cloud storage in a way that is easy to access.

## Prerequisites
This document assumes that you have installed and can run the ledger exporter, and that you have familiarity with its CLI and configuration. If not, please refer to the [Link to Admin Guide]

## Goal

> The goal of the ledger exporter is to build an easy-to-use tool to export Stellar network ledger data to cloud storage.

Reviewer: ```suggestion
The goal of the ledger exporter is to build an easy-to-use tool to export Stellar network ledger data to a configurable remote data store, such as cloud blob storage.
```

 - Use cloud storage optimally
 - Minimize network usage to export
 - Make it easy and fast to search for a specific ledger or ledger range

## Architecture
To achieve its goals, the ledger exporter uses the following architecture, which consists of the 3 main components:
- Captive-core [Link to existing documentation on captive-core basics] to extract raw transaction metadata from the Stellar Network
- Export manager to bundles and organizes the ledgers to get them ready for export

> - The cloud storage plugin writes to the cloud storage. This is specific to the type of cloud storage, GCS in this case.

Reviewer: Maybe mention the `DataStore` interface/package as the aspect that provides pluggability? and `GCSDataStore` as one implementation.

> [Insert Architecture diagram here]

Reviewer: here's a diagram from the eng talk that may be reusable, still lines up with current day

![galexie](https://github.com/stellar/go/assets/10703845/73558e42-bcb4-4120-af23-9ba380494e28)

## Data Format
- Ledger exporter uses a compact and efficient data format called XDR (External Data Representation), which is a compact binary format. The captive-core emits data in this format and the data structure is referred to as `LedgerCloseMeta`. The exporter bundle multiple LedgerCloseMeta's into a single object using a custom XDR struct called LedgerCloseMetaBatch which is defined in [Stellar-exporter.x](https://github.com/stellar/go/blob/master/xdr/Stellar-exporter.x).

- The metadata for the same batch is also stored with for each exported object. Supported metadata is defined in [metadata.go](https://github.com/stellar/go/blob/master/support/datastore/metadata.go). 

- Objects are compressed before uploading using the zstd (zstandard) compression algorithm to optimize network usage and storage needs.

## Data Storage (GCS)

> - The source for the cloud storage plugin is in the [support](https://github.com/stellar/go/tree/master/support/datastore) package. Any new storage plugin must implement the interface defined in [datastore.go](https://github.com/stellar/go/blob/master/support/datastore/datastore.go). 

Reviewer: ```suggestion
- An example implementation of `DataStore` for GCS, Google Cloud Storage. This plugin is located in the [support](https://github.com/stellar/go/tree/master/support/datastore) package. 
```

- The ledger exporter currently implements the interface only for Google Cloud Storage (GCS). The [GCS plugin](https://github.com/stellar/go/blob/master/support/datastore/gcs_datastore.go) uses GCS-specific behaviors like conditional puts, automatic retry, metadata, and CRC checksum.

## Build, Run and Test using Docker
The Dockerfile contains all the necessary dependencies (e.g., Stellar-core) required to run the ledger exporter. 
- Build: To build the Docker container, use the provided [Makefile](https://github.com/stellar/go/exp/services/ledgerexporter/Makefile). Simply run make `make docker-build` to build a new container after making any changes.

- Run: For instructions on running the Docker container, refer to the Admin Guide [Link to Admin Guide].

- Test: To test the Docker container, refer to the [docker-test](https://github.com/stellar/go/blob/master/exp/services/ledgerexporter/Makefile) command for an example of how to use the [GCS emulator](https://github.com/fsouza/fake-gcs-server) for local testing. 

## Adding support for a new storage type

> To add support for a new storage type (e.g. AWS S3), follow these steps:

Reviewer: ```suggestion
Support for different data storage types are encapsulated as 'plugins', which are implementation of `DataStore` interface in a go package. To add a data storage plugin based on a new storage type (e.g. AWS S3), follow these steps:
```

> - Implement the interface defined in [datastore.go](https://github.com/stellar/go/blob/master/support/datastore/datastore.go).

Reviewer: ```suggestion
- A data storage plugin must implement the [DataStore](https://github.com/stellar/go/blob/master/support/datastore/datastore.go) interface.
```

# Ledger Exporter Developer Guide
The ledger exporter is a tool to export Stellar network transaction data to cloud storage in a way that is easy to access.

## Prerequisites
This document assumes that you have installed and can run the ledger exporter, and that you have familiarity with its CLI and configuration. If not, please refer to the [Installation Guide](./README.md).

## Goal
The goal of the ledger exporter is to build an easy-to-use tool to export Stellar network ledger data to a configurable remote data store, such as cloud blob storage.
 - Use cloud storage optimally
 - Minimize network usage to export
 - Make it easy and fast to search for a specific ledger or ledger range

## Architecture
To achieve its goals, the ledger exporter uses the following architecture, which consists of the 3 main components:
- Captive-core to extract raw transaction metadata from the Stellar Network.
- Export manager to bundles and organizes the ledgers to get them ready for export.
- The cloud storage plugin writes to the cloud storage. This is specific to the type of cloud storage, GCS in this case.

> ![ledgerexporter-architecture](https://github.com/urvisavla/go/assets/30014929/8fcc4733-6e35-4de7-ad06-0a27515364cd)

Reviewer 2: Should this asset be added to the go monorepo instead of your personal one?

Author: done

## Data Format

> - Ledger exporter uses a compact and efficient data format called XDR (External Data Representation), which is a compact binary format. The captive-core emits data in this format and the data structure is referred to as `LedgerCloseMeta`. The exporter bundle multiple LedgerCloseMeta's into a single object using a custom XDR struct called LedgerCloseMetaBatch which is defined in [Stellar-exporter.x](https://github.com/stellar/go/blob/master/xdr/Stellar-exporter.x).

Reviewer 3: ```suggestion
- Ledger exporter uses a compact and efficient data format called [XDR](https://developers.stellar.org/docs/learn/encyclopedia/data-format/xdr) (External Data Representation), which is a compact binary format. A Stellar Captive Core instance emits data in this format and the data structure is referred to as `LedgerCloseMeta`. The exporter bundles multiple `LedgerCloseMeta`'s into a single object using a custom XDR structure called `LedgerCloseMetaBatch` which is defined in [Stellar-exporter.x](https://github.com/stellar/go/blob/master/xdr/Stellar-exporter.x).
```

> - The metadata for the same batch is also stored with for each exported object. Supported metadata is defined in [metadata.go](https://github.com/stellar/go/blob/master/support/datastore/metadata.go). 

Reviewer 3: ```suggestion
- The metadata for the same batch is also stored alongside each exported object. Supported metadata is defined in [metadata.go](https://github.com/stellar/go/blob/master/support/datastore/metadata.go). 
```

> - Objects are compressed before uploading using the zstd (zstandard) compression algorithm to optimize network usage and storage needs.

Reviewer 3: ```suggestion
- Objects are compressed before uploading using the [zstd](http://facebook.github.io/zstd/) (zstandard) compression algorithm to optimize network usage and storage needs.
```
